### DISCUSSION TITLE
Testing framework for Doodle

### DISCUSSION THREAD

**Discussion by Krzyjan** [nacular/doodle#40] [Q&A] [answered]
Can someone recommend a testing framework for Doodle UI? It should be able to run automated tests from Gradle. I am looking for something similar to testfx(http://testfx.github.io/TestFX/) used for JavaFX.

**Comment by pusolito** [comment, marked as answer]
In general you should be able to write unit style tests with mocks (I recommend [Mockk](https://github.com/mockk/mockk)). There are some basic examples of this in the [Calculator tutorial](https://nacular.github.io/doodle-tutorials/docs/calculator) ([repo](https://github.com/nacular/doodle-tutorials/blob/master/Calculator/src/jvmTest/kotlin/io/nacular/doodle/examples/CalculatorTests.kt)).

You can fully mock your `Application` instance as well if you inject everything; which is the recommendation.

Unfortunately, there isn't a framework (as far as I know) that offers something like the fluent API of the JavaFX one you shared.

**Reply by Krzyjan** [reply]
Thanks for a prompt response.  I will use your CalculatorTest for now. Perhaps someone will come up with a framework simulating user interactions eventually :-)
